python/reg_interface/scripts/sca.py

```diff
@@ -64,51 +45,99 @@ def main():
+            printRed('if your firmware file is a *.msc file use the --fwFileMCS argument')
             return
 
-        ftype = sys.argv[4]
-        filename = sys.argv[5]
         program_fpga(ohMask,ftype,filename)
 
-    elif instructions == 'test1':
-        test1()
+    #elif instructions == 'test1':
+    #    test1()
         
-    elif instructions == 'test2':
-        test2()
+    #elif instructions == 'test2':
+    #    test2()
 
     elif instructions == 'compare-mcs-bit':
-        if len(sys.argv) < 5:
-            print("Usage: sca.py local compare-mcs-bit <mcs_filename> <bit_filename>")
+        if ((fwFileBit is None) or (fwFileMCS is None)):
+            print("Usage: sca.py --cardName=<cardName> --cmd='compare-mcs-bit' --fwFileMCS=<mcs_filename> --fwFileBit=<bit_filename>")
             return
-        mcsFilename = sys.argv[3]
-        bitFilename = sys.argv[4]
-        compare_mcs_bit(mcsFilename, bitFilename)
+        from reg_utils.reg_interface.arm.program_fpga import compare_mcs_bit
+        compare_mcs_bit(fwFileMCS, fwFileBit)
 
     elif instructions == 'gpio-set-direction':
-        if len(sys.argv) < 4:
-            print('Usage: sca.py local gpio-set-direction <direction-mask>')
+        if gpioValue is None:
+            print('Usage: sca.py --cardName=<cardName> --cmd="gpio-set-direction" --gpioValue=<direction-mask>')
             print('direction-mask is a 32 bit number where each bit represents a GPIO channel -- if a given bit is high it means that this GPIO channel will be set to OUTPUT mode, and otherwise it will be set to INPUT mode')
             return
-        directionMask = parseInt(sys.argv[3])
-        gpio.set_direction(ohMask,directionMask)
+        gpio.set_direction(ohMask,gpioValue)
 
     elif instructions == 'gpio-set-output':
-        if len(sys.argv) < 4:
-            print('Usage: sca.py local gpio-set-output <output-data>')
+        if gpioValue is None:
+            print('Usage: sca.py --cardName=<cardName> --cmd="gpio-set-output" --gpioValue=<output-data>')
             print('output-data is a 32 bit number representing the 32 GPIO channels state')
             return
-        outputData = parseInt(sys.argv[3])
-        gpio.set_output(ohMask,outputData)
+        gpio.set_output(ohMask,gpioValue)
 
     elif instructions == 'gpio-read-input':
         gpio.read_input(ohMask)
  
 if __name__ == '__main__':
-    main()
+    supportedCmds = [
+            'compare-mcs-bit',
+            'h',
+            'hh',
+            'fpga-id',
+            'gpio-read-input',
+            'gpio-set-direction',
+            'gpio-set-output',
+            'program-fpga',
+            'r',
+            'sysmon',
+            #'test1',
+            #'test2',
+            ]
+    from optparse import OptionParser
+    import argparse
+    parser = argparse.ArgumentParser(description='Arguments to supply to sca.py')
```

Comment:
I think my main point about using the `argparse` library was missed:
Current (primary) usage should be preserved
* Required args (well, arguments are always technically required) should still be positional, e.g.,
```bash
sca.py eagle33 r 0x3
sca.py eagle33 fpga-id 0x3
```
* Additional (command dependent) options should have their sensibility tested as is currently done
```bash
sca.py eagle33 program-fpga 0x3 --fwFileBit
```
Ideally, there would be a nice way to cobble together the usage options using some `argparse` option groups, but I haven't dug into this at all really

Reply:
> Required args (well, arguments are always technically required) should still be positional, e.g.,

I disagree.  Having positional arguments has caused numerous bugs due to differences between privately maintained `sca.py` on the `texas` account and the one in this repository.  These bugs are explicitly because the fact they are positional.  The goal here is to drop this positional requirement.  

This would come with an update to the documentation; which I can do myself when it is deployed.

Reply:
Actually, perhaps for the last point [`sub_commands`](https://docs.python.org/2/library/argparse.html#sub-commands) would be ideal for this, although it would require reordering `ohMask` and `cmd`

```python
    parser = argparse.ArgumentParser()
    group = parser.add_mutually_exclusive_group()
    group.add_argument("-v", "--verbose", action="store_true")
    group.add_argument("-q", "--quiet",   action="store_true")

    ## Positional arguments
    parser.add_argument("cardName", help="Card name", type=str)
    parser.add_argument("ohMask",   help="OH link mask", type=int)
    subparsers = parser.add_subparsers(help='sub-command help')

    # create the parser for the "standard" commands
    parser_a = subparsers.add_parser('h',      help='h help')
    parser_b = subparsers.add_parser('hh',     help='hh help')
    parser_c = subparsers.add_parser('r',      help='r help')
    parser_d = subparsers.add_parser('sysmon', help='sysmon help')

    # create the parser for the "program-fpga" command
    parser_e = subparsers.add_parser('program-fpga', help='a help')
    fpgs_group = parser_e.add_mutually_exclusive_group()
    # still have to parse that at least one was provided, as a mutually exclusive option can't be `required=True`
    fpgs_group.add_argument('--fwBitFile', type=str, help='FW image in bit format')
    fpgs_group.add_argument('--fwMcsFile', type=str, help='FW image in mcs format')

    ## Options
    parser.add_argument("-d",    help="debug", action='store_true')

    args = parser.parse_args()
```
This would print the help message:
```bash
usage: argtester.py [-h] [-v | -q] [-d]
                    cardName ohMask {h,hh,r,sysmon,program-fpga} ...

positional arguments:
  cardName              Card name
  ohMask                OH link mask
  {h,hh,r,sysmon,program-fpga}
                        sub-command help
    h                   h help
    hh                  hh help
    r                   r help
    sysmon              sysmon help
    program-fpga        a help

optional arguments:
  -h, --help            show this help message and exit
  -v, --verbose
  -q, --quiet
  -d                    debug
```

and the specific help message for `program-fpga`
```bash
usage: argtester.py cardName ohMask program-fpga [-h]
                                                 [--fwBitFile FWBITFILE | --fwMcsFile FWMCSFILE]

optional arguments:
  -h, --help            show this help message and exit
  --fwBitFile FWBITFILE
                        FW image in bit format
  --fwMcsFile FWMCSFILE
                        FW image in mcs format
```

Reply:
I'll have a look and try to implement.